### PR TITLE
Added exercise 5 firmware update implementation

### PR DESCRIPTION
**Pull request by malw4**

### CODE REVIEW

File: docs/academy/exercise5.md [hunk 1]
```diff
@@ -1,20 +1,446 @@
...
+
+If no errors occur, the update process follows this pattern:
+
+1. **Downloading** `state 1` & `update result 0`
+2. **Downloaded** `state 2` & `update result 0`
+3. **Updating** `state 3` & `update result 0`
+4. **Updated** `state 0` & `update result 1`
+
+!!! important "Update successful?"
+    Does the **State** `/5/*/3` report `0` and the **Update Results** `/5/*/5` report `1`? Congratulations! You've successfully updated the firmware of your device. 🎉
```
Comment (JZimnol) [inline docs/academy/exercise5.md:446]:
```suggestion
    Does the **State** `/5/*/3` report `0` and the **Update Result** `/5/*/5` report `1`? Congratulations! You've successfully updated the firmware of your device 🎉
```

File: docs/academy/gradproject.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+## Useful Resources
+* [Getting Started guides](https://iotdevzone.avsystem.com/docs/LwM2M_Client/Getting_started/) containing detailed documentation on using different IoT development kits
+* [AVSystem Discord](https://discord.avsystem.com/) for support and for submitting your graduation project
+
+
+!!! important "Provide your feedback"
+
+    We're constantly working on improving the LwM2M Academy. Please share with us your feedback about this module so we can create an even better learning experience.
+
+    [Feedback form](https://forms.gle/UAFLJs9LJocAeger9){: .md-button .md-button--big }
```
Comment (JZimnol) [inline docs/academy/gradproject.md:20]: The feedback form does not include the Module 5 Exercise

**Review by JZimnol — COMMENTED** [review]
@malw4 reviewed. Remember to NOT merge this before the newest `pico-client` is released.

File: docs/academy/exercise5.md [hunk 3]
```diff
@@ -1,20 +1,446 @@
-# Graduation Project
+# Exercise 5: Implement Firmware Update
+In this exercise, we implement Object 5 - Firmware Update.  It utilizes the pico_fota_bootloader to swap the flash partitions after downloading the appropriate binary file from Coiote IoT DM.
 
-Congratulations on your progress in the LwM2M Academy! As you approach graduation, one final task awaits: bringing your IoT idea to life using the LwM2M standard.
+## Prerequisites
 
-Your graduation project involves building an application that utilizes the LwM2M standard and the {{ coiote_long_name }}. Whether you choose the Raspberry Pi Pico W or any other IoT device, your task is to effectively gather sensor data and establish a cloud connection.
+* A Raspberry Pi Pico W board with a USB cable
+* A LM35 temperature sensor
```
[the remark this answers is not in the document]
Comment (malw4) [reply docs/academy/exercise5.md:7]: sorry, copy and paste gone wrong

File: docs/academy/exercise5.md [hunk 4]
```diff
@@ -1,20 +1,446 @@
...
+    .reset = fw_reset,
+    .perform_upgrade = fw_perform_upgrade
+};
+int fw_update_install(anjay_t *anjay) {
+    anjay_fw_update_initial_state_t state = { 0 };
+
+    if (pfb_is_after_firmware_update()) {
+        state.result = ANJAY_FW_UPDATE_INITIAL_SUCCESS;
+        avs_log(fw_update, INFO, "Running on a new firmware");
+    }
```
[the remark this answers is not in the document]
Comment (malw4) [reply docs/academy/exercise5.md:238]: then please update the github version

File: docs/academy/exercise5.md [hunk 5]
```diff
@@ -1,20 +1,446 @@
...
+If the Firmware Update is scheduled successfully, the device starts **downloading** the firmware at the next practical opportunity. The actual firmware **update** starts once the integrity and authenticity of the firmware image has been validated by the LwM2M Client.
+
+![Upgrading](images/in-progress-basic.png)
+
+Once executed successfully, the status in the **Update list** panel changes to `Success`.
+
+![Successful update](images/success.png)
+
+!!! note
+    While the device is updating its firmware, it will deregister and reboot using the new firmware. This process may time several minutes.
```
[the remark this answers is not in the document]
Comment (malw4) [reply docs/academy/exercise5.md:426]: look at this https://iotdevzone.avsystem.com/docs/Coiote_IoT_DM/Single_update/#download-upgrade-process. There is also this note

File: docs/academy/exercise5.md [hunk 6]
```diff
@@ -1,20 +1,398 @@
...
+## Implement Firmware Update
 
-**Best of luck with this final step. We can't wait to see how you apply your acquired knowledge to create a compelling end-to-end application.**
+Anjay comes with a built-in Firmware Update module, which simplifies FOTA implementation for the user. Let’s dive into the code and discuss its most important fragments.
 
-## Useful Resources
-* [Getting Started guides](https://iotdevzone.avsystem.com/docs/LwM2M_Client/Getting_started/) containing detailed documentation on using different IoT development kits
-* [AVSystem Discord](https://discord.avsystem.com/) for support and for submitting your graduation project
+!!! Note
+    This part only describes functions that are in the code in **Anjay-pico-client/firmware_update** folder. The user doesn’t have to modify the code.
```
Comment (JZimnol) [inline docs/academy/exercise5.md:17]:
```suggestion
    This part only describes functions that are in the code in **Anjay-pico-client/firmware_update** directory. The user doesn’t have to modify the code.
```

File: docs/academy/exercise5.md [hunk 7]
```diff
@@ -1,20 +1,398 @@
...
+
+## Flash alignment
+
+Flash APIs require that the length of data to write will be a multiple of 256 bytes, so we need to enforce that by additional buffering. For this in the firmware_update directory there are two more files:
+
+- flash_aligned_writer.c
+
+<p style="text-align: center;">flash_aligned_writer.c</p>
+``` c
+    #include <assert.h>
```
Comment (JZimnol) [inline docs/academy/exercise5.md:208]: You've indented the code, but not the whole section. It renders badly. Applies for other code snippets.

File: docs/academy/module5.md [hunk 8]
```diff
@@ -73,6 +73,11 @@ Learn more about the Firmware Update Object on the [OMA LwM2M Object and Resourc
     | `update result 10`  | **Firmware update cancelled** |
     | `update result 11`  | **Firmware update deferred** |
 
+
+## What is Bootloader and why do I need this here?
```
Comment (JZimnol) [inline docs/academy/module5.md:77]: No no no, this place is `z dupy` for that in my opinion. THis hsould be in Exercise 5, before `Implement Firmware Update` for example.

File: docs/academy/exercise5.md [hunk 9]
```diff
@@ -443,4 +395,4 @@ If no errors occur, the update process follows this pattern:
 4. **Updated** `state 0` & `update result 1`
 
 !!! important "Update successful?"
-    Does the **State** `/5/*/3` report `0` and the **Update Results** `/5/*/5` report `1`? Congratulations! You've successfully updated the firmware of your device. 🎉
\ No newline at end of file
+    Does the **State** `/5/*/3` report `0` and the **Update Result** `/5/*/5` report `1`? Congratulations! You've successfully updated the firmware of your device. 🎉
```
Comment (JZimnol) [inline docs/academy/exercise5.md:398]: It is single instance, so no `*`.

**Review by JZimnol — COMMENTED** [review]
@malw4 re-reviewed.

File: docs/academy/exercise5.md [hunk 10]
```diff
@@ -203,106 +208,106 @@ Flash APIs require that the length of data to write will be a multiple of 256 by
...
-        assert(writer_cb);
-
-        out_writer->batch_buf = batch_buf;
-        out_writer->batch_buf_max_len_bytes = batch_buf_max_len_bytes;
-        out_writer->batch_buf_len_bytes = 0;
-        out_writer->write_offset_bytes = 0;
-        out_writer->writer_cb = writer_cb;
-    }
+    <p style="text-align: center;">flash_aligned_writer.c</p>
+    ``` c
```
Comment (JZimnol) [inline docs/academy/exercise5.md:212]: The whole code is still indented not the way it should be. The "code section" is OK itself, the code inside is indented unnecessarily.

File: docs/academy/exercise5.md [hunk 11]
```diff
@@ -203,106 +208,106 @@ Flash APIs require that the length of data to write will be a multiple of 256 by
...
+        #include <assert.h>
+        #include <stddef.h>
+        #include <stdint.h>
+
+        #include <avsystem/commons/avs_utils.h>
+
+        #include "flash_aligned_writer.h"
+
+        void flash_aligned_writer_new(uint8_t *batch_buf,
+                                    size_t batch_buf_max_len_bytes,
```
Comment (JZimnol) [inline docs/academy/exercise5.md:222]:
```suggestion
        void flash_aligned_writer_new(uint8_t *batch_buf,
                                      size_t batch_buf_max_len_bytes,
```

And everywhere else, the code is not aligned after copy-paste.

**Review by JZimnol — COMMENTED** [review]
The code snippets in `firmware_update.c` are still not up to date or are incorrect.

**Review by JZimnol — APPROVED** [review]
@malw4  approving in advance, but please make any necessary changes discussed privately

**Pull request merged**